cmd/autoexport.go

```diff
@@ -50,3 +45,64 @@ func AutoExport(c *cli.Context) {
 	logrus.Info("Start listening for new Docker events...")
 	b.ListenEventsForExported(ctx, promNetwork)
 }
+
+// Thread-safe collection of context.CancelFunc
+type cancellableCollection struct {
+	mutex sync.RWMutex
+	funcs map[string]context.CancelFunc
+}
+
+func newCancellableCollection() cancellableCollection {
+	return cancellableCollection{
+		mutex: sync.RWMutex{},
+		funcs: make(map[string]context.CancelFunc, 0),
+	}
+}
+
+func (c cancellableCollection) cancel(k string) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	var f func()
+	var ok bool
+
+	if f, ok = c.funcs[k]; ok {
+		f()
+		delete(c.funcs, k)
+	}
+
+	return ok
+}
+
+func (c cancellableCollection) add(k string, ctx context.Context) context.Context {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	ctx, c.funcs[k] = context.WithCancel(ctx)
+	return ctx
+}
+
+func (c cancellableCollection) remove(k string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.funcs[k]; ok {
+		delete(c.funcs, k)
+	}
+}
+
+// @TODO: implement true back-off retry
+func retry(times uint, interval time.Duration, f func() error) error {
```

Comment:
is this fn still used ?

Reply:
Yep it's gonna be use to implement a sort of retry to start/stop exporters ;)